File: C# .Net/JDI UI Framework/JDI/Web/Selenium/Elements/Complex/Selector.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Epam.JDI.Core.Interfaces.Complex;
using OpenQA.Selenium;
using static Epam.JDI.Core.Settings.JDISettings;

namespace Epam.JDI.Web.Selenium.Elements.Complex
{
    public class Selector : Selector<IConvertible>, ISelector
    {
        public Selector() { }
        public Selector(By optionsNamesLocatorTemplate, By allOptionsNamesLocator) 
            : base(optionsNamesLocatorTemplate, allOptionsNamesLocator) { }
    }
    public class Selector<TEnum> : BaseSelector<TEnum>, ISelector<TEnum>
        where TEnum : IConvertible
    {
        protected Selector() : this(null, null) { }
        public Selector(By optionsNamesLocatorTemplate, By allOptionsNamesLocator)
            : base(optionsNamesLocatorTemplate, allOptionsNamesLocator)
        {
            SelectedNameAction = (s, name) => ((Selector<TEnum>) s).SelectedAction(this).Equals(name);
            SelectedNumAction = (s, num) => ((Selector<TEnum>)s).SelectedIndexAction(this) == num;
            GetValueAction = s => ((Selector<TEnum>)s).Selected();
        }
        public void Select(string name) {
            Actions.Select(name, (s, n) => SelectNameAction(this, name));
        }

        public void Select(TEnum enumType) {
            Select(enumType.ToString());
        }

        public void Select(int index) {
            Actions.Select(index, (s, n) => SelectNumAction(this, index));
        }

        public string Selected() {
            return Actions.Selected(s => SelectedAction(this));
        }

        public int SelectedIndex() {
            return Actions.SelectedIndex(s => SelectedIndexAction(this));
        }

        public Func<Selector<TEnum>, string> SelectedAction = s =>
            {
                if (s.AllLabels != null)
                    return s.Selected(s.AllLabels.WebElements);
                if (s.Locator.ToString().Contains("{0}"))
                    throw Exception(
                        "Can't get Selected options. Override getSelectedAction or place locator to <select> tag");
                var els = s.WebAvatar.SearchAll().WebElements;
                return s.Selected(els.Count == 1 ? s.Selector.Options.ToList() : els);
            };

        public string Selected(IList<IWebElement> els) {
            var element = els.FirstOrDefault(el => SelectedElementAction(this, el));
            if (element == null)
                throw Exception("No elements selected. Override getSelectedAction or place locator to <select> tag");
            return element.Text;
        }

        public Func<Selector<TEnum>, int> SelectedIndexAction = s =>
        {
            if (s.AllLabels != null)
                return s.SelectedIndex(s.AllLabels.WebElements);
            if (s.Locator.ToString().Contains("{0}"))
                throw Exception(
                    "Can't get Selected options. Override getSelectedAction or place locator to <select> tag");
            var els = s.WebAvatar.SearchAll().WebElements;
            return s.SelectedIndex(els.Count == 1 ? s.Selector.Options.ToList() : els);
        };

        private int SelectedIndex(List<IWebElement> els) {
            var index = els.FindIndex(el => SelectedElementAction(this, el)) + 1;
            if (index == 0)
                throw Exception("No elements selected. Override getSelectedAction or place locator to <select> tag");
            return index;
        }

    }
}

File: Java/JDI/jdi-uitest-web/src/main/java/com/epam/jdi/uitests/web/selenium/elements/complex/Selector.java
package com.epam.jdi.uitests.web.selenium.elements.complex;
/*
 * Copyright 2004-2016 EPAM Systems
 *
 * This file is part of JDI project.
 *
 * JDI is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * JDI is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; 
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with JDI. If not, see <http://www.gnu.org/licenses/>.
 */


import com.epam.jdi.uitests.core.interfaces.complex.ISelector;
import org.openqa.selenium.By;
import org.openqa.selenium.WebElement;

import java.util.List;

import static com.epam.commons.EnumUtils.getEnumValue;
import static com.epam.commons.LinqUtils.first;
import static com.epam.commons.LinqUtils.firstIndex;
import static com.epam.jdi.uitests.core.settings.JDISettings.exception;

/**
 * Created by roman.i on 03.10.2014.
 */

public class Selector<TEnum extends Enum> extends BaseSelector<TEnum> implements ISelector<TEnum> {
    public Selector() {
        super();
    }

    public Selector(By optionsNamesLocatorTemplate) {
        super(optionsNamesLocatorTemplate);
    }

    public Selector(By optionsNamesLocatorTemplate, By allOptionsNamesLocator) {
        super(optionsNamesLocatorTemplate, allOptionsNamesLocator);
    }

    public final void select(String name) {
        actions.select(name, this::selectAction);
    }

    public final void select(TEnum name) {
        select(getEnumValue(name));
    }

    public final void select(int num) {
        actions.select(num, this::selectAction);
    }

    public final String getSelected() {
        return actions.getSelected(this::getSelectedAction);
    }

    public final int getSelectedIndex() {
        return actions.getSelectedIndex(this::getSelectedIndexAction);
    }

    protected final boolean isSelectedAction(String name) {
        return getSelectedAction().equals(name);
    }

    protected final boolean isSelectedAction(int num) {
        return getSelectedIndexAction() == num;
    }

    protected String getValueAction() {
        return getSelected();
    }

    protected String getSelectedAction() {
        return getSelected(getElements());
    }

    private String getSelected(List<WebElement> els) {
        WebElement element = first(els, this::isSelectedAction);
        if (element == null)
            throw exception("No elements selected. Override getSelectedAction or place locator to <select> tag");
        return element.getText();
    }

    protected int getSelectedIndexAction() {
        return getSelectedIndex(getElements());
    }

    private int getSelectedIndex(List<WebElement> els) {
        int num = firstIndex(els, this::isSelectedAction) + 1;
        if (num == 0)
            throw exception("No elements selected. Override getSelectedAction or place locator to <select> tag");
        return num;
    }
}